crates/ide_assists/src/handlers/convert_into_to_from.rs

```diff
@@ -0,0 +1,295 @@
+    let impl_ = ctx.find_node_at_offset::<ast::Impl>()?;
+    let src_type = impl_.self_ty()?;
+    let ast_trait = impl_.trait_()?;
+
+    impl_is_an_into(&impl_, &ctx)?;
+
+    let dest_type = match &ast_trait {
+        ast::Type::PathType(path) => {
+            path.path()?.segment()?.generic_arg_list()?.generic_args().next()?
+        }
+        _ => return None,
+    };
+
+    let src_type_path_segment =
+        src_type.syntax().descendants().find_map(ast::Path::cast)?.segment()?;
+
+    let into_fn = impl_.assoc_item_list()?.assoc_items().find_map(|item| {
+        if let ast::AssocItem::Fn(f) = item {
+            if f.name()?.text() == "into" {
+                return Some(f);
+            }
+        };
+        None
+    })?;
+
+    let into_fn_name = into_fn.name()?;
+    let into_fn_params = into_fn.param_list()?;
+    let into_fn_return = into_fn.ret_type()?;
+
+    let selfs = into_fn
+        .body()?
+        .syntax()
+        .descendants()
+        .filter_map(ast::NameRef::cast)
+        .filter(|name| name.text() == "self" || name.text() == "Self");
+
+    acc.add(
+        AssistId("convert_into_to_from", AssistKind::RefactorRewrite),
+        "Convert Into to From",
+        impl_.syntax().text_range(),
+        |builder| {
+            builder.replace(src_type.syntax().text_range(), dest_type.to_string());
+            builder.replace(ast_trait.syntax().text_range(), format!("From<{}>", src_type));
+            builder.replace(into_fn_return.syntax().text_range(), "-> Self");
+            builder.replace(
+                into_fn_params.syntax().text_range(),
+                format!("(val: {})", src_type.to_string()),
+            );
+            builder.replace(into_fn_name.syntax().text_range(), "from");
+
+            let src_type_without_generics = match src_type_path_segment.generic_arg_list() {
+                Some(generic_params) => src_type_path_segment
+                    .replace_children(single_node(generic_params.syntax().clone()), iter::empty()),
+
+                None => src_type_path_segment,
+            };
+
+            for s in selfs {
+                match s.text().as_ref() {
+                    "self" => builder.replace(s.syntax().text_range(), "val"),
+                    "Self" => builder
+                        .replace(s.syntax().text_range(), src_type_without_generics.to_string()),
+                    _ => {}
+                }
+            }
+        },
+    )
+}
+
+fn impl_is_an_into(impl_: &ast::Impl, ctx: &AssistContext) -> Option<()> {
+    let trait_ = resolve_target_trait(&ctx.sema, &impl_)?;
+    let module = ctx.sema.scope(impl_.syntax()).module()?;
+    let krate = module.krate();
+    let into_trait = FamousDefs(&ctx.sema, Some(krate)).core_convert_Into()?;
+
+    (into_trait == trait_).then(|| ())
+}
+
+fn single_node(element: impl Into<SyntaxElement>) -> RangeInclusive<SyntaxElement> {
+    let element = element.into();
+    element.clone()..=element
+}
```

Comment:
I don't think this function is needed since its only used one, inlining it seems better. Also see https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#helper-functions